src/Compilers/Shared/DesktopBuildClient.cs

```diff
@@ -313,6 +317,41 @@ protected override int RunLocalCompilation(string[] arguments, BuildPaths buildP
             }
         }
 
+        // Protected for testing
+        protected static bool TryConnectToNamedPipeWithSpinWait(NamedPipeClientStream pipeStream,
+                                                                int timeoutMs,
+                                                                CancellationToken cancellationToken)
+        {
+            // .NET 4.5 implementation of NamedPipeStream.Connect busy waits the entire time.
+            // Work around is to spin wait.
+            const int maxWaitInterval = 50;
+            int elapsed = 0;
+            var sw = new SpinWait();
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                int waitTime = Math.Min(timeoutMs - elapsed, maxWaitInterval);
+                if (timeoutMs != Timeout.Infinite && waitTime <= 0)
+                {
+                    return false;
+                }
+
+                try
+                {
+                    pipeStream.Connect(waitTime);
+                    break;
+                }
+                catch (Exception e) when (e is IOException || e is TimeoutException)
+                {
+                    // Ignore timeout
+                }
+                elapsed += waitTime;
+                sw.SpinOnce();
```

Comment:
I don't think addressing this scenario is important. Time-outs in all systems that I know of are "suggestions" rather than rules. There's also no guarantee that this thread could be preempted by the processor for an arbitrary long period of time, causing the timeout to extend far beyond the stated time.

I view the timeout as a "reasonable effort" request, rather than a requirement.